mqtt/mqtt-broker/src/sys_topic.rs

```diff
@@ -0,0 +1,399 @@
+
+const STATE_CHANGE_QOS: proto::QoS = proto::QoS::AtLeastOnce;
+
+pub enum StateChange<'a> {
+    Subscriptions(&'a ClientId, Option<Vec<&'a str>>),
+    Connections(Vec<&'a ClientId>),
+    Sessions(Vec<&'a ClientId>),
+}
+
+impl<'a> StateChange<'a> {
+    pub fn new_subscription(client_id: &'a ClientId, session: &'a Session) -> Self {
+        let subscriptions = session
+            .subscriptions()
+            .map(|s| s.keys().map(String::as_str).collect());
+
+        Self::Subscriptions(client_id, subscriptions)
+    }
+
+    pub fn clear_subscriptions(client_id: &'a ClientId) -> Self {
+        Self::Subscriptions(client_id, None)
+    }
+
+    pub fn new_connection(sessions: &'a HashMap<ClientId, Session>) -> Self {
+        let connections = sessions
+            .iter()
+            .filter_map(|(client_id, session)| match session {
+                Session::Transient(_) => Some(client_id),
+                Session::Persistent(_) => Some(client_id),
+                _ => None,
+            })
+            .collect();
+
+        Self::Connections(connections)
+    }
+
+    pub fn new_session(sessions: &'a HashMap<ClientId, Session>) -> Self {
+        let sessions = sessions
+            .iter()
+            .filter_map(|(client_id, session)| match session {
+                Session::Transient(_) => Some(client_id),
+                Session::Persistent(_) => Some(client_id),
+                Session::Offline(_) => Some(client_id),
+                _ => None,
+            })
+            .collect();
+
+        Self::Sessions(sessions)
+    }
+}
+
+impl<'a> TryFrom<StateChange<'a>> for proto::Publication {
+    type Error = Error;
+
+    fn try_from(state: StateChange<'a>) -> Result<Self, Error> {
+        Ok(match state {
+            StateChange::Subscriptions(client_id, subscriptions) => {
+                let payload = subscriptions
+                    .map(|subscriptions| serde_json::to_string(&subscriptions))
+                    .transpose()?
+                    .map_or_else(bytes::Bytes::new, |json| json.into());
+
+                proto::Publication {
+                    topic_name: format!("$edgehub/subscriptions/{}", client_id),
+                    qos: STATE_CHANGE_QOS,
+                    retain: true,
+                    payload,
+                }
+            }
+            StateChange::Connections(connections) => proto::Publication {
+                topic_name: "$edgehub/connected".to_owned(),
+                qos: STATE_CHANGE_QOS,
+                retain: true,
+                payload: serde_json::to_string(&connections)?.into(),
+            },
+            StateChange::Sessions(sessions) => proto::Publication {
+                topic_name: "$edgehub/sessions".to_owned(),
+                qos: STATE_CHANGE_QOS,
+                retain: true,
+                payload: serde_json::to_string(&sessions)?.into(),
+            },
+        })
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
```

Comment:
I would suggest going forward to make **unit tests** more granular. Here in every test you are testing both `new_***` and `try_into` together. Instead, I would split it into separate tests. This is not a big deal, and this is fine as well, but just as a recommendation going forward.